packages/style-engine/class-wp-style-engine-css-rules-store.php

```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * WP_Style_Engine_CSS_Rules_Store
+ *
+ * A store for WP_Style_Engine_CSS_Rule objects.
+ *
+ * @package Gutenberg
+ */
+
+if ( class_exists( 'WP_Style_Engine_CSS_Rules_Store' ) ) {
+	return;
+}
+
+/**
+ * Holds, sanitizes, processes and prints CSS declarations for the style engine.
+ *
+ * @access private
+ */
+class WP_Style_Engine_CSS_Rules_Store {
+
+	/**
+	 * An array of named WP_Style_Engine_CSS_Rules_Store objects.
+	 *
+	 * @static
+	 *
+	 * @var WP_Style_Engine_CSS_Rules_Store[]
+	 */
+	protected static $stores = array();
+
+	/**
+	 * An array of CSS Rules objects assigned to the store.
+	 *
+	 * @var WP_Style_Engine_CSS_Rule[]
+	 */
+	protected $rules = array();
+
+	/**
+	 * Get an instance of the store.
+	 *
+	 * @param string $store_name The name of the store.
+	 *
+	 * @return WP_Style_Engine_CSS_Rules_Store
+	 */
+	public static function get_store( $store_name ) {
+		if ( ! isset( static::$stores[ $store_name ] ) ) {
+			static::$stores[ $store_name ] = new WP_Style_Engine_CSS_Rules_Store( $store_name );
+		}
+		return static::$stores[ $store_name ];
+	}
+	/**
+	 * Get an array of all rules.
+	 *
+	 * @return WP_Style_Engine_CSS_Rule[]
+	 */
+	public function get_all_rules() {
+		return $this->rules;
+	}
+
+	/**
+	 * Get a WP_Style_Engine_CSS_Rule object by its selector.
+	 *
+	 * @param string $selector The CSS selector.
+	 */
+	public function get_rule( $selector ) {
```

Comment:
Unless I'm missing something, we'll also need to add rules to stores from around Gutenberg.

When we ultimately want to enqueue a store's rules, then we'd want to retrieve them all at once.

Reply:
Ah apologies, I think the `get_rule` threw me and I expected a corresponding `set_rule`. 

The references penny didn't drop. 😄 

It did eventually when I tried to write tests: https://github.com/WordPress/gutenberg/pull/42408/files#diff-676e332de83a58d26bc1513e708615a18ebddc8edcc23d804272e9f1feeaafd4

So just to confirm we'd be doing something like this:

```php
// Deal with incoming css rules from around Gutenberg.
$store_rule         = $new_pizza_store->get_rule( $selector );
$css_declarations   = new WP_Style_Engine_CSS_Declarations( $pizza_declarations );
$store_rule->set_declarations( array( $css_declarations ) );

// Somewhere else... eventually.
$new_pizza_store->get_all_rules();

// Build the stylesheet from the rules...
```

Reply:
That's a good point about getters vs setters and potential confusion. Perhaps using a separate setter could make the API a bit more explicit?

One of the things for us to keep in mind as we explore the style engine API is what sort of behaviour (and naming) we expect in WordPress APIs in general. So, for example, if the `WP_Block_Type_Registry`, `WP_Block_Styles_Registry`, and `WP_Block_Patterns_Registry` classes have separate `register` and `get_registered` methods (and `get_registered` returns `null` if the item is not registered), then consumers of this class might expect comparable behaviour / methods.

Reply:
I think I found a use case for an explicit `set_rule`: https://github.com/WordPress/gutenberg/pull/42408/files#r920830003

Aside from the idea that it might be handy for people like me who don't like to think too hard 😆

Reply:
You're right, `get_rule()` is probably not the right name for this method.
What if we called it `add_rule` and also added a `remove_rule` method? 🤔

Reply:
`add_rule` and `remove_rule` work for me!